SamBumgardner/csc-303-platformer
# Added Powerup Mushroom to the class game

**Overview**

This pull request serves to implement a powerup that alters the players state when collected.

**Details**
This pull request adds the classic mushroom powerup to our game. The mushroom is animated to move back and forth on the screen, and is affected by gravity. When it is collected, the player is given 2 units of health. Observe how the player is killed after a single hit from the sentry or Goomba enemy without the powerup, but is able to withstand 2 hits from either enemy after the powerup is collected. 

I also implemented a timer-based invincibility system that grants the player 2 seconds of invincibility immediately after taking damage. This was implemented to avoid the player taking damage over consecutive frames, and dying instantaneously when in contact with the Goomba. Observe how the player can remain in contact with the Goomba for 2-3 seconds before being killed when the mushroom is collected.

There is also a fireflower powerup in the level. If collected, this powerup will give the player 3 units of health total. Please ignore, as it is functional, but incomplete, and is not the focus of this pull request.

**In the future**
I would like to add some sort of visual effect to show when the player is in the "invincible" state, most likely by using a tweener of some sort to alter the players color or opacity. The same effect can be recycled to show when the powerup is collected if desired.

I would also like to place the mushroom inside of an item block, rather than spawning it directly into the play area on start.

Ultimately, I would like to have the fire flower be functional also, granting the ability to launch fireballs that damage/kill enemies. However, this is lowest on the list of priorities, and will most likely not be completed before class's end.

**To test**

* Esure you have Haxe and HaxeFlixel installed on your computer.
* Installing HaxeFlixel should also automatically install OpenFl and Lime.
* After HaxeFlixel's installation has finished, open a new command prompt and run `haxelib install flixel-addons` to install an additional library of extra HaxeFlixel features.
* Clone or download this repository to your computer.
* Open a command prompt in the newly created \csc-303-platformer folder.
* Run the command `haxelib run lime test neko` to build and run the executable.
* The executable is located in \csc-303-platformer\export\windows\neko\bin.
* To run in debug mode, run the command `haxlib run lime test -debug neko`. The debug console can be accessed with the backquote key.
